source/IAccessibleHandler/__init__.py

```diff
@@ -239,6 +244,99 @@
+
+
+def _getStatesSetFromIAccessibleStates(
+		IAccessibleStates: int
+) -> Set[controlTypes.State]:
+	return set(
+		IAccessibleStatesToNVDAStates[IAState]
+		for IAState in IAccessibleStatesToNVDAStates.keys()
+		if IAState & IAccessibleStates
+	)
+
+
+def getStatesSetFromIAccessible2States(IAccessible2States: int) -> Set[State]:
+	return set(
+		IAccessible2StatesToNVDAStates[IA2State]
+		for IA2State in IAccessible2StatesToNVDAStates.keys()
+		if IA2State & IAccessible2States
+	)
+
+
+def getStatesSetFromIAccessibleAttrs(attrs: "textInfos.ControlField") -> Set[State]:
+	# States are serialized (in XML) with an attribute per state.
+	# The value for the state is used in the attribute name.
+	# The attribute value is always 1.
+	# EG IAccessible::state_40="1"
+	IAccessibleStateAttrName = 'IAccessible::state_{}'
+	return set(
+		IAccessibleStatesToNVDAStates[IAState]
+		for IAState in IAccessibleStatesToNVDAStates.keys()
+		if int(attrs.get(IAccessibleStateAttrName.format(IAState), 0))
+	)
+
+
+def getStatesSetFromIAccessible2Attrs(attrs: "textInfos.ControlField") -> Set[State]:
+	# States are serialized (in XML) with an attribute per state.
+	# The value for the state is used in the attribute name.
+	# The attribute value is always 1.
+	# EG IAccessible2::state_40="1"
+	IAccessible2StateAttrName = 'IAccessible2::state_{}'
+	return set(
+		IAccessible2StatesToNVDAStates[IA2State]
+		for IA2State in IAccessible2StatesToNVDAStates.keys()
+		if int(attrs.get(IAccessible2StateAttrName.format(IA2State), 0))
+	)
+
+
+def calculateNvdaRole(IARole: int, IA2States: int) -> Role:
+	NVDARole = IAccessibleRolesToNVDARoles.get(IARole, Role.UNKNOWN)
+	# Don't report indeterminate progress bars as "half-checked"
+	# L{State.HALFCHECKED} maps from oleacc.STATE_SYSTEM_MIXED
+	# which has the same value as oleacc.STATE_SYSTEM_INDETERMINATE
+	# L{State.INDETERMINATE} is not mapped directly from any IA or IA2 state
+	# A progress bar that can not convey progress, only activity, is a busy indicator.
+	if NVDARole == Role.PROGRESSBAR and oleacc.STATE_SYSTEM_INDETERMINATE & IA2States:
+		return Role.BUSY_INDICATOR
+	return NVDARole
+
+
+def calculateNvdaStates(IARole: int, IA2States: int) -> Set[State]:
+	roles = IAccessibleRolesToNVDARoles.get(IARole, Role.UNKNOWN)
+	states = _getStatesSetFromIAccessibleStates(IA2States)
+	# Don't report indeterminate progress bars as "half-checked"
+	# L{State.HALFCHECKED} maps from oleacc.STATE_SYSTEM_MIXED
+	# which has the same value as oleacc.STATE_SYSTEM_INDETERMINATE
+	# L{State.INDETERMINATE} is not mapped directly from any IA or IA2 state
+	# A progress bar that can not convey progress, only activity, is a busy indicator.
+	if roles == Role.PROGRESSBAR and states.intersection({State.HALFCHECKED, State.INDETERMINATE}):
+		states.discard(State.HALFCHECKED)
+		states.add(State.INDETERMINATE)
+		return states
+	return states
+
+
+def NVDARoleFromAttr(accRole: Union[int, str, None]) -> Role:
+	if accRole.isdigit():

```

Comment:
`None.isdigit()` and `int.isdigit()` raises an AttributeError

Reply:
Note that this code is copied verbatim from usages. The type information is fuzzy, it was guessed based on usage.

Reply:
I've fixed this by making it accept an optional string, handle None / empty string explicitly, and assert the type is string. There is one usage that was potentially calling with an `int` value `0`. This will now call with `None`.